src/wp-includes/rest-api/endpoints/class-wp-rest-abilities-run-controller.php

```diff
@@ -0,0 +1,241 @@
+<?php
+/**
+ * REST API run controller for Abilities API.
+ *
+ * @package WordPress
+ * @subpackage Abilities_API
+ * @since 6.9.0
+ */
+
+declare( strict_types = 1 );
+
+/**
+ * Core controller used to execute abilities via the REST API.
+ *
+ * @since 6.9.0
+ *
+ * @see WP_REST_Controller
+ */
+class WP_REST_Abilities_Run_Controller extends WP_REST_Controller {
+
+	/**
+	 * REST API namespace.
+	 *
+	 * @since 6.9.0
+	 * @var string
+	 */
+	protected $namespace = 'wp/v2';
+
+	/**
+	 * REST API base route.
+	 *
+	 * @since 6.9.0
+	 * @var string
+	 */
+	protected $rest_base = 'abilities';
+
+	/**
+	 * Registers the routes for ability execution.
+	 *
+	 * @since 6.9.0
+	 *
+	 * @see register_rest_route()
+	 */
+	public function register_routes(): void {
+		register_rest_route(
+			$this->namespace,
+			'/' . $this->rest_base . '/(?P<name>[a-zA-Z0-9\-\/]+?)/run',
+			array(
+				'args'   => array(
+					'name' => array(
+						'description' => __( 'Unique identifier for the ability.' ),
+						'type'        => 'string',
+						'pattern'     => '^[a-zA-Z0-9\-\/]+$',
+					),
+				),
+
+				// TODO: We register ALLMETHODS because at route registration time, we don't know which abilities
+				// exist or their annotations (`destructive`, `idempotent`, `readonly`). This is due to WordPress
+				// load order - routes are registered early, before plugins have registered their abilities.
+				// This approach works but could be improved with lazy route registration or a different
+				// architecture that allows type-specific routes after abilities are registered.
+				// This was the same issue that we ended up seeing with the Feature API.
+				array(
+					'methods'             => WP_REST_Server::ALLMETHODS,
+					'callback'            => array( $this, 'execute_ability' ),
+					'permission_callback' => array( $this, 'check_ability_permissions' ),
+					'args'                => $this->get_run_args(),
+				),
+				'schema' => array( $this, 'get_run_schema' ),
+			)
+		);
+	}
+
+	/**
+	 * Executes an ability.
+	 *
+	 * @since 6.9.0
+	 *
+	 * @param WP_REST_Request<array<string, mixed>> $request Full details about the request.
+	 * @return WP_REST_Response|WP_Error Response object on success, or WP_Error object on failure.
+	 */
+	public function execute_ability( $request ) {
+		$ability = wp_get_ability( $request->get_param( 'name' ) );
+		if ( ! $ability ) {
+			return new WP_Error(
+				'rest_ability_not_found',
+				__( 'Ability not found.' ),
+				array( 'status' => 404 )
+			);
+		}
+
+		$input  = $this->get_input_from_request( $request );
+		$result = $ability->execute( $input );
+		if ( is_wp_error( $result ) ) {
+			if ( 'ability_invalid_input' === $result->get_error_code() ) {
+				$result->add_data( array( 'status' => 400 ) );
```

Comment:
I might only add this if the `status` data is not present? That way folks can use more specific status codes when appropriate.

Reply:
This was refactored in https://github.com/WordPress/wordpress-develop/pull/9410/commits/bde0aa6e6479b4d2f5913c1b9f58c6b296ddb91f based on the feedback in https://github.com/WordPress/wordpress-develop/pull/9410#discussion_r2445966990. 

I can follow-up with enhancements there. For now, during the permission check we execute input validation seperately, if it fails, we send 400 HTTP status, which means that the request wasn't properly formed.